### PR TITLE
feat(annotation-thread): Truncate the message in Comment / AnnotationActivity

### PR DESCRIPTION
**Pull request by jankowiakdawid**
We want to truncate long messages to only first 7 lines. When any massage is longer we'll truncate it and show button to show more.

#### Message truncated as it would look like in IE
Screenshot 2022-10-06 at 22 59 14

#### Message truncated as it would look like in modern browsers
Screenshot 2022-10-06 at 23 01 40

#### Full message
Screenshot 2022-10-06 at 22 59 30

### CODE REVIEW

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (karelee7) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss:30]: Should we store this px amount in a variable? Or make a note about how we got 140px exactly?

File: src/elements/content-sidebar/activity-feed/common/activity-message/__tests__/ActivityMessage.test.js
Comment (karelee7) [inline src/elements/content-sidebar/activity-feed/common/activity-message/__tests__/ActivityMessage.test.js:140]: Should we also add a test for when it shouldn't be truncated?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js
Comment (bfoxx1906) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:182]: So it looks like shouldTruncate doesn't determine whether or not a comment is truncated it determines if truncation should be possible. shouldTruncate could be true but truncation may never happen because the messages aren't long enough right? If this is the case then It might be more intuitive to name this something along the lines of **allowTruncation**.

Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:182]: Exactly! 💯 Great suggestion! I double checked in designs, and I've noticed they use different name for this `Collapse` so I think I'll go with full rename for the component.
But will definitely change this prop to `allowCollapsing`

File: src/elements/content-sidebar/activity-feed/common/activity-message/TruncatableMessage.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/TruncatableMessage.js:19]: We're currently duplicating this value in the stylesheet. Can we share the value between the component and stylesheet? If not, should we apply this one style rule via inline styles, instead?

File: src/elements/content-sidebar/activity-feed/common/activity-message/TruncatableMessage.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/TruncatableMessage.js:34]: Minor, but we should be able to use `&&` here, instead.

File: src/elements/content-sidebar/activity-feed/common/activity-message/__tests__/ActivityMessage.test.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/__tests__/ActivityMessage.test.js:147]: Minor, but we typically prefer `expect(wrapper.exists('TruncatableMessage')).toBe(true)`

File: src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js:19]: Definitely!

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js
Comment (tjuanitas) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:182]: this is minor but `allow` isn't a common prefix we use for booleans. Would `canCollapse` suffice?

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (tjuanitas) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss:27]: typo
```suggestion
    /* webkit-line-clamp only works in combination with -webkit-box */
```

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (tjuanitas) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss:35]: I'm not familiar with these or `display: -webkit-box`. what does this do? does this work in other browsers (Firefox, Safari)?

File: src/elements/content-sidebar/activity-feed/common/activity-message/TruncatableMessage.js
Comment (tjuanitas) [inline src/elements/content-sidebar/activity-feed/common/activity-message/TruncatableMessage.js:13]: I think this is being renamed but in case not:
```suggestion
    const [shouldTruncate, setShouldTruncate] = React.useState(false);
```

File: src/elements/content-sidebar/activity-feed/common/activity-message/messages.js
Comment (tjuanitas) [inline src/elements/content-sidebar/activity-feed/common/activity-message/messages.js:21]: minor: the `id`s for the messages above have a `Message` suffix

`id: 'be.contentSidebar.activityFeed.common.seeMoreMessage',`
`id: 'be.contentSidebar.activityFeed.common.seeLessMessage',`

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss]: It allows to define after how many lines text should be cut, and adds `...` at the end
It is standardized for legacy support

https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp
https://caniuse.com/css-line-clamp

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:182]: Is this desired? It will immediately reflect in the live Preview experience for long comments.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplies.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplies.js:78]: Minor, but we typically try to keep our props, destructuring, etc. alphabetized when possible.

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.js:8]: Alphabetize?

File: src/styles/constants/_layout.scss
Comment (jstoffan) [inline src/styles/constants/_layout.scss:27]: I understand we created these variables to share them between the component/stylesheet. However, these are global values. Do we intend to use them elsewhere? What happens if other experiences have different font sizes and/or line heights? It may be preferable to move the values into the JS component and use them via inline styles. What do you think?

File: src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js:19]: Minor, but can we move `type` to the end?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:182]: Without it the top-level comments won't be collapsible. Is there a check that I can put inside so that it will be true only for Threaded Comments? like:
```suggestion
                                        allowCollapse={hasReplies}
```
?

I'll bring it to the team tomorrow.

File: src/styles/constants/_layout.scss
Comment (jankowiakdawid) [reply src/styles/constants/_layout.scss:27]: I need some time to think this through. I don't like to use inline styles.

Comment (tjuanitas) [reply src/styles/constants/_layout.scss:27]: I would also hesitate to put them in this file/directory. at minimum they would need a `bdl-` prefix.

since they only have one usage at the moment, is it okay to have them as local variables in `ActivityMessage.scss`?

File: src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js:19]: Can we have lint rule for this?

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.js:8]: Can we have lint rule for this?

File: src/styles/constants/_layout.scss
Comment (jankowiakdawid) [reply src/styles/constants/_layout.scss:27]: I wanted them in `ActivityMessage.scss` but the `:export {}` syntax didn't worked. Then I found that we have build script for generating js files with sass variables.
I think I'll go with `styles`.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:182]: @kkuliczkowski-box suggested using features for this, that way it would be enabled for Canvas only for now.
Similarly to this
https://github.com/box/box-ui-elements/blob/master/src/elements/content-sidebar/ActivitySidebar.js#L892

I found that I can use https://github.com/box/box-ui-elements/blob/master/src/elements/common/feature-checking/FeatureContext.js

I decided to put that in single place in `Comment.js`. What do you think?
I have implementation ready, but still some tests are failing before I commit.

Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js]: here's the commit https://github.com/box/box-ui-elements/pull/3079/commits/f807afc79df5c9a09992de997043b61474462183

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.js
Comment (tjuanitas) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.js]: Yeah we have an internal ticket to investigate updating our prettier config. It's sort of waiting on a third party plugin and would need a review from security as well. But it's something that definitely should be added to the repo.

File: src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js
Comment (jstoffan) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js]: Sounds fantastic. The issue we've had in the past with new rules like that is that there's lots of non-conforming code that needs to be updated.

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js:11]: We can remove this logic, if this is the case, since new features do not have to support IE11.

File: src/elements/content-sidebar/activity-feed/comment/Comment.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/comment/Comment.js:258]: Should this be handled within `ActivityMessage` so annotation messages also get the same experience?

File: src/elements/content-sidebar/activity-feed/comment/Comment.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/comment/Comment.js:72]: Is this needed?

Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/comment/Comment.js:72]: No

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js
Comment (JChan106) [inline src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js:9]: Do we need to define children in the props? I think React components come with children built into the props

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js
Comment (JChan106) [inline src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js:32]: Do we need to use inline styling if we use `isCollapsed` in the classname?

File: src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js
Comment (JChan106) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ToggleMoreLessButton.js:15]: Can we type this with the arrow syntax `() => void` to be more explicit to what the props and return value are

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js
Comment (JChan106) [inline src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js:14]: Should we export default inline or at the end of files?

File: src/elements/content-sidebar/activity-feed/comment/Comment.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/comment/Comment.js:258]: I mean, I don't know. It can. Probably would be better for the users.

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js:32]: Actually, we don't need it anymore. It was for backward compatibility with IE11. But as @jstoffan pointed out it's not required.

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessage.js:9]: We need. https://flow.org/en/docs/react/children/

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss:36]: Minor, but this can be simplified:

```
.bcs-ActivityMessage-toggleMoreLess,
.bcs-ActivityMessage-translate {
    &.btn-plain {
        // ... styles
    }
}

File: src/elements/content-sidebar/activity-feed/common/activity-message/__tests__/ActivityMessage.test.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/__tests__/ActivityMessage.test.js:4]: Should we be testing the default export by wrapping it in a FeaturesProvider, instead? It's what other components will use.

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss:36]: I didn't knew exact syntax for that, but I knew you'll tell me 😉

File: src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/ActivityMessage.scss:44]: No. If we won't have this transparent border then when user will interact with the button, button will move due to appearing border. My solution to that was to always have border but keep it invisible. This lead to more white space around those button which I then needed to subtract.

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessageToggle.js
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessageToggle.js:19]: Minor, but it's best to pass this in from the `ActivityMessage` component. SUIT naming asks for each component to have its own namespace. Apologies for missing this on prior reviews.

**Review by jstoffan — DISMISSED** [review]

File: src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessageToggle.js
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessageToggle.js:19]: I will read more on SUIT naming. For now, I leave it as it is, because I see similar naming in other components inside `./activity-message`. Can it be that all the names are in context of `ActivityMessage`  which is a single place of styling for all those small components that were just extracted from `ActivityMessage` but aren't standalone on their own?

Comment (jstoffan) [reply src/elements/content-sidebar/activity-feed/common/activity-message/CollapsableMessageToggle.js:19]: We may have made similar mistakes in other components, but that doesn't mean it's correct. Each component is supposed to have its own top-level name and hyphens are used to indicate descendent elements within a component. Since this is a separate file, it should have a separate class name. Either way, this issue alone isn't enough to block the merge.

**Review by jstoffan — APPROVED** [review]

**Pull request merged**
